### PR TITLE
Have a simpler "getting started" notebook

### PR DESCRIPTION
**Pull request by pierreablin**
This PR adds a more basic "getting started" notebook that covers only the very basic use of ott: compute OT between two datasets and differentiate through it.

### CODE REVIEW

**Comment by pierreablin** [comment]
Any comments on what I should add / remove are more than welcome :)

File: docs/tutorials/notebooks/basic_ot_between_datasets.ipynb [hunk 1]
```diff
@@ -0,0 +1,376 @@
+{
```
Comment (marcocuturi) [inline docs/tutorials/notebooks/basic_ot_between_datasets.ipynb:1]: <p>maybe link directly Sinkhorn with the solver module?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/296/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (marcocuturi) [inline docs/tutorials/notebooks/basic_ot_between_datasets.ipynb:1]: <p>Since this is a low-level tutorial, maybe worth adding a few words on what a geometry is. </p><p> </p><p>Essentially this geometry is just two (or one) point clouds compared with a <code>SqEuclidean</code> cost between points here. Maybe this can be discussed by setting <code>cost_fn</code> to None (the default) and mention this defaults to <code>SqEuclidean</code> (linked in doc).</p><p> </p><p>Maybe also mention that <code>epsilon</code> regularization in Sinkhorn is also stored in the geometry (and set by default to 1/20th of mean cost value). This can come later when running Sinkhorn (e.g. a comment  "Wait? Sinkhorn? but where's <code>epsilon</code> ? ) </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/296/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (marcocuturi) [inline docs/tutorials/notebooks/basic_ot_between_datasets.ipynb:1]: <p>Maybe we could describe the two interfaces, either <code>.solve</code> or create a solver with all sorts of <code>.Sinkhorn()(...)</code>. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/296/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (marcocuturi) [inline docs/tutorials/notebooks/basic_ot_between_datasets.ipynb:1]: <p>a colormap would be nice, to insist this is a coupling matrix (values &lt; 1), and so we expect row-sum / column-sums to sum to 1/n et 1/m, and total coupling matrix sum to 1.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/296/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (marcocuturi) [inline docs/tutorials/notebooks/basic_ot_between_datasets.ipynb:1]: <p>i think we can find something nicer. I'll try to put the code for figure 4 in our <a href="https://arxiv.org/abs/2302.04065" target="_blank">paper</a></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/296/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

Comment (bamos) [inline docs/tutorials/notebooks/basic_ot_between_datasets.ipynb:1]: <p>It could be nice to also plot the components of this gradient on top of the data in the scatterplot above as another way of illustrating how it pushes the measures closer together</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/296/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (pierreablin) [reply docs/tutorials/notebooks/basic_ot_between_datasets.ipynb:1]: <p>I feel like this is a bit too much for a tutorial. I can instead use the Sinkhorn class if this is a more recommended way to do ?</p><p> </p>

Comment (pierreablin) [reply docs/tutorials/notebooks/basic_ot_between_datasets.ipynb:1]: <p>To me there is a big tradeoff to find in terms of simplicity of the experiment and of the plotting : I'd really avoid having 15 lines of matplotlib code that does something prettier but less clear to the beginner. I have in mind the sklearn examples which use very rudimentary matplotlib (https://scikit-learn.org/stable/auto_examples/cluster/plot_bisect_kmeans.html#sphx-glr-auto-examples-cluster-plot-bisect-kmeans-py) but like that are cristal clear.</p>

Comment (pierreablin) [reply docs/tutorials/notebooks/basic_ot_between_datasets.ipynb:1]: <p>good idea :) I added this.</p>

**Comment by pierreablin** [comment]
Thanks for the comments ! I implemented most of the suggested changes.

File: docs/tutorials/notebooks/basic_ot_between_datasets.ipynb (hunk 1 above)
Comment (marcocuturi) [reply docs/tutorials/notebooks/basic_ot_between_datasets.ipynb:1]: <p>With so many colors selected randomly (not ideal for color blind people) and points scattered all around (markers are the same for x / y), I have trouble visualizing anything in that plot. I agree that the code is more compact (definitely a +) but I feel we're losing in terms of visualization. Maybe something more intermediate?</p>

File: docs/tutorials/notebooks/basic_ot_between_datasets.ipynb [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
```
Comment (marcocuturi) [inline docs/tutorials/notebooks/basic_ot_between_datasets.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #3.    plt.quiver(y[:, 0], y[:, 1], -gradients[:, 0], -gradients[:, 0])
</pre><p>,1 for second, here gradients are on the y=x axis</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/296/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

Comment (marcocuturi) [inline docs/tutorials/notebooks/basic_ot_between_datasets.ipynb:1]: <p>standard squared-Euclidean distance</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/296/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (marcocuturi) [inline docs/tutorials/notebooks/basic_ot_between_datasets.ipynb:1]: <p>i think it's more natural to differentiate <code>reg_ot_cost</code> using Danskin (differentiating <code>primal_cost</code> triggers implicit differentiation)  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/296/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (marcocuturi) [inline docs/tutorials/notebooks/basic_ot_between_datasets.ipynb:1]: <p>i guess, if i must nitpick, <code>gradients</code> should be <code>gradient</code>,since this is the gradient for the entire point cloud that's shown   </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/296/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

File: docs/index.rst [hunk 3]
```diff
@@ -82,7 +82,7 @@ Packages
     :maxdepth: 1
     :caption: Examples
 
-    Getting Started <tutorials/notebooks/point_clouds>
+    Getting Started <tutorials/notebooks/basic_ot_between_datasets>
```
Comment (marcocuturi) [inline docs/index.rst:85]: this means the `point_cloud` note book is removed completely, but I think it had its own use to highlight other functionalities (such as defining alternative cost functions)

**Review by marcocuturi — COMMENTED** [review]
Thanks Pierre! a few comments

File: docs/tutorials/notebooks/basic_ot_between_datasets.ipynb [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
```
Comment (marcocuturi) [inline docs/tutorials/notebooks/basic_ot_between_datasets.ipynb:1]: <ul><li>the convention to have target before source is a bit unusual, maybe the other way round (does not change much, because it's symmetric, but agrees better with the usual convention.</li><li>given the goal to have a very compact NB, not sure this wrapper is really needed. wouldn't it be easier to have for instance</li></ul><p><code>grad_fn = jax.grad(lambda x, y : sinkhorn.solve(PointCloud(x, y)).reg_ot_cost</code>  in that sense it does not really matter if we return both gradients.</p><ul><li>Let's avoid using capital letters for variable names (naming convention of google python style)</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/296/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (pierreablin) [reply docs/tutorials/notebooks/basic_ot_between_datasets.ipynb:1]: <p>Thanks ! </p><p> </p><p>About the wrapper, I feel like it is much easier to read like this, right? the lambda etc might be a bit too much for a new user?</p>

Comment (pierreablin) [reply docs/tutorials/notebooks/basic_ot_between_datasets.ipynb:1]: <p>Also I reverted the names to <code>x</code> and <code>y</code>  like in the pointsclouds notebook, let me know if <code>x_source</code> / <code>x_target</code> are better names.   </p>

Comment (marcocuturi) [reply docs/tutorials/notebooks/basic_ot_between_datasets.ipynb:1]: <p>maybe lambda is a bit heavy, but </p><ul><li><code>cost_from_source</code> is not a very convincing function name (it's too long). </li><li>I don't think the closure on <code>y</code> is needed. It's ok to keep the <code>def ...</code>  if you prefer to avoid the lambda, but adding <code>y</code> to the list of arguments won't make any difference (<code>jax.grad</code> has <code>arg_nums</code> to 0 by default) and I think it's more simple. </li></ul>

File: docs/tutorials/notebooks/basic_ot_between_datasets.ipynb [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
```
Comment (marcocuturi) [inline docs/tutorials/notebooks/basic_ot_between_datasets.ipynb:1]: <p>i am nitpicking, but this is needed, because this is page 1 of the toolbox.</p><p> </p><p>The original introductory nb describes what a point cloud geometry is, and this discussion has been completely removed here. I am not sure we should save space on this, because it's quite important (see e.g. https://github.com/ott-jax/ott/discussions/314)</p><p> </p><p>In particular, since <code>geometry</code> also refers to the module name and base class <code>Geometry</code>,it might be a bit confusing to call your <code>PointCloud</code> object as <code>geometry</code>.Maybe <code>geom</code> is more neutral, or <code>pc_geom</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/296/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (pierreablin) [reply docs/tutorials/notebooks/basic_ot_between_datasets.ipynb:1]: What specific information in the pointcloud tutorial is missing here? Unless you want to revert to instantiating a LinearProblem, everything is pretty much there. 

I fixed the name.

File: docs/tutorials/notebooks/basic_ot_between_datasets.ipynb (hunk 2 above)
Comment (marcocuturi) [reply docs/tutorials/notebooks/basic_ot_between_datasets.ipynb:1]: <p>maybe "The Sinkhorn algorithm <strong>uses</strong>"</p>

File: docs/tutorials/notebooks/basic_ot_between_datasets.ipynb [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
```
Comment (marcocuturi) [inline docs/tutorials/notebooks/basic_ot_between_datasets.ipynb:1]: <p>"that is a way to measure distances between points" -&gt; "that is a way to measure a ground cost between points"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/296/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (marcocuturi) [inline docs/tutorials/notebooks/basic_ot_between_datasets.ipynb:1]: <p>thanks for writing this extra paragraph, it's very cool!</p><p> </p><p>"of different masses" -&gt; "of different total masses."</p><p> </p><p>I also feel it would be great to have links, either to other tutorials or maybe to the functions, rather than plain text. An extra line on neural approaches could be nice.</p><p> </p><p>BTW, Low-rank Sinkhorn is not so much about using a low-rank approx of the cost matrix (this can be used), but rather implement a low-rank <strong>constraint </strong>on the coupling.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/296/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

Comment (pierreablin) [reply docs/tutorials/notebooks/basic_ot_between_datasets.ipynb:1]: <p>I have now added these new approaches. Not sure where the neural approaches fit here, I can also put it in the "numerical tricks to make ot scale", wdyt?</p>
